### PR TITLE
Solve #376: clashing in case of more classes with the same name

### PR DESCRIPTION
**Pull request by fabioformosa**
# What does the Feature do

This PR avoids clashing in validation check, if more classes has the same name.
* With `typeAlias` attribute in `@prop()` is possibile to disambiguate name of types used as nested object.
* With `typeAlias` attribute in `GetModelForClassOption` is possible to disambiguate models with the same class name.

Please, look at:
* Issue #376 to get more details about what this feature solves
* the README.md changes to see examples of use of `typeAlias`
* the new test case `tests/typeAlias.test.ts`

## Related Issues

- #376

### CODE REVIEW

**Comment by hasezoey** [comment]
sorry, but could you re-target it to the r6/master (on my fork)?
-> to be in version 6.0.0

**Comment by fabioformosa** [comment]
> sorry, but could you re-target it to the r6/master (on my fork)?
> -> to be in version 6.0.0
Yes, I can. But I'm using the current version 5.9.0, and I'm very interested to have this feature in ver 5.
Could you merge, please? Have you permission to do it?

> 
> PS: please read the CONTRIBUTING file and how to structure commits for PR's, thanks (i mean especially to squash as needed, thanks)

Where's the CONTRIBUTING file?

**Comment by hasezoey** [comment]
> Where's the CONTRIBUTING file?

to your right, when you start an pull request, [otherwise here](https://github.com/szokodiakos/typegoose/blob/master/.github/CONTRIBUTING.md)

---

> Have you permission to merge and to publish in npm registry?

that is the reason why v6 is currently not on the offical npm & offical repo(this one here), because only @szokodiakos can give the rights to this

but v6 is release under `@hasezoey/typegoose@next` and in #365 is the changelog

---

here are some things my fork has, but the offical repo (this one) dosnt: hasezoey#1

**Comment by fabioformosa** [comment]
> @szokodiakos can give the rights to this

I'm confused!
From npmjs.com, I see that @Ben305 has rights to publish.

Hi @Ben305, please could you clarify me some points:
* Are you currently the maintainer of this project?
* Could you accept my PR?
* Could you publish a new version?

**Comment by hasezoey** [comment]
> Are you currently the maintainer of this project?

yes he is a maintainer (/collaborator), but currently dosnt do any code changes & Accepting PR's (look at the still open ones, and how "old" they are), only currently accepted changes are documentation changes, hes not often doing much here in the last time

**Comment by hasezoey** [comment]
btw i am now looking through your pr a bit, and it seems you modified many documentation changes (Readme, etc) that are "not related" to the actual solving of "clash in case of more classes with same name", so could you spread it in multiple PR'S?

**Comment by fabioformosa** [comment]
> btw i am now looking through your pr a bit, and it seems you modified many documentation changes (Readme, etc) that are "not related" to the actual solving of "clash in case of more classes with same name", so could you spread it in multiple PR'S?

Unfortunately some format option caused diff also in rows that I didn't edit.
I will try to understand if is a format option of my IDE or a formatter included in the project (like prettier).

**Comment by fabioformosa** [comment]
> yes he is a maintainer (/collaborator), but currently dosnt do any code changes & Accepting PR's (look at the still open ones, and how "old" they are), only currently accepted changes are documentation changes, hes not often doing much here in the last time
> 
> PS: look at #99

I've just pushed @szokodiakos to grant you maintainer permission as asked by Ben in #99

**Comment by hasezoey** [comment]
> I've just pushed @szokodiakos to grant you maintainer permission as asked by Ben in #99

yes i saw

File: test/models/type-alias-event2.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import * as mongoose from 'mongoose';
+
+import { prop, Typegoose } from '../../src/typegoose';
+
+export class Event extends Typegoose {
+  // constructor(fieldTwo: string, fieldOne?: string) {
```
Comment (hasezoey) [inline test/models/type-alias-event2.ts]: why is this code commented-out and still in here?

**Review by hasezoey — CHANGES_REQUESTED** [review]
please add a test with "typeAlias" for https://github.com/szokodiakos/typegoose/blob/aaec1bfdb50ab1effee65b1af1dd0f0241024fc1/src/utils.ts#L81-L88

for reference
https://github.com/szokodiakos/typegoose/blob/master/test/tests/getClassForDocument.test.ts

File: test/models/type-alias-event2.ts (hunk 1 above)
Comment (fabioformosa) [reply test/models/type-alias-event2.ts]: Oh sorry, I'm going to remove it

**Comment by fabioformosa** [comment]
> please add a test with "typeAlias" for
I've already written it. See `test/tests/typeAlias.test.ts`

**Comment by hasezoey** [comment]
> I've already written it. See test/tests/typeAlias.test.ts

nope, cant find it

**Comment by fabioformosa** [comment]
> > I've already written it. See test/tests/typeAlias.test.ts
> 
> nope, cant find it

It's true. I'm going to understand this function and I'll write a test.

**Comment by hasezoey** [comment]
>  I'm going to understand this function and I'll write a test.

trying to help: this function gets the class from the the supplied document, if it can be found in the class "array" (not in 6.0.0 its a map)

File: test/tests/getClassForDocument.test.ts [hunk 2]
```diff
@@ -45,12 +46,18 @@ export function suite() {
     expect(userReflectedType).to.not.equals(CarType);
   });
 
+  it(`should return correct class type for document also using typeAlias`, async () => {
+    const event = new EventModel();
```
Comment (hasezoey) [inline test/tests/getClassForDocument.test.ts]: but are you *sure* that it is just this model, when only having one named "Event"?
-> please add to be *sure* and dummy Event and make it different properties, thanks

Comment (fabioformosa) [reply test/tests/getClassForDocument.test.ts]: Yes, I'm sure. As you asked, I've added another model named Event with different properties and typeAlias.

**Comment by hasezoey** [comment]
does [this](hasezoey.github.io/typegoose/guides/advanced/models-with-same-name/) | f73cca004682e85a63896353c9dcfc1311c2b464 somehow relate what this pr tries to achieve?

**Comment by fabioformosa** [comment]
hi @hasezoey, It seems to me that solves different models (extending typegoose) with same class name.
But it doens't solve the case of subtypes (not extending typegoose class) with same class name. Do you agree?
my PR has some troubles, I'm fixing them.

**Comment by hasezoey** [comment]
the listed commit, is from version 6.0.0, so "extending typegoose" is deprecated, and is not used

PS: if you got redirected to a wrong URL, sorry, my bad, fixed it

**Comment by fabioformosa** [comment]
Could you apply your commit to this project that reproduces my error with class name conflict?
https://github.com/fabioformosa/typegoose-validation-error
Thank you

**Comment by hasezoey** [comment]
(didnt test or run it yet, but wanted to say some things:)

the commit i listed here, makes differing models, with "same name", but only if you add an option for "collection" or a "customName" into `@modelOptions`

**Comment by fabioformosa** [comment]
Ok so I'm sure that it cannot work in my case, because I have a conflict between two classes with same name but one ([bar](https://github.com/fabioformosa/typegoose-validation-error/blob/master/src/model/bar/bar.ts)) is a model  (with `@modelOption`) and the other one ([bar](https://github.com/fabioformosa/typegoose-validation-error/blob/master/src/model/foo/bar.ts)) is a subtype (with no collection) of [foo](https://github.com/fabioformosa/typegoose-validation-error/tree/master/src/model/foo).
See

**Comment by hasezoey** [comment]
i just run your thing, changed it to v6 syntax, and no build errors, should i look at something different?

**Comment by fabioformosa** [comment]
Have you launched `npm test`? do they pass?

**Comment by hasezoey** [comment]
i ran npm build

---

npm test:
- one test passes
- one fails: "fieldTwo": Path `fieldTwo` is required
otherwise it runs fine

PS: because there is no option to change mongodb connections, i had to improvise, that could result in errors

**Comment by fabioformosa** [comment]
> * one fails: "fieldTwo": Path `fieldTwo` is required
>   otherwise it runs fine

So this proves that your versione doesn't fix the case reproduced here https://github.com/fabioformosa/typegoose-validation-error

**Comment by hasezoey** [comment]
as you can see, i deleted my comment, because i did something wrong, will report again

**Comment by hasezoey** [comment]
i ran npm build

---

npm test:
- both passes

**Comment by fabioformosa** [comment]
> give me write permission to your test repo, and i will give the change to you (in a different branch, to compare)

I've just given it to you

**Comment by hasezoey** [comment]
pushed to `typegoosev6`

**Comment by fabioformosa** [comment]
I got, but `npm install` raise error. Have you pushed your package in public npm repo?

**Comment by hasezoey** [comment]
no https://www.npmjs.com/package/@hasezoey/typegoose/v/6.0.0-25

**Comment by fabioformosa** [comment]
Ok npm problem solved.

I have a question: Why did you add `getModelForClass(Bar)` to the nested [bar](https://github.com/fabioformosa/typegoose-validation-error/blob/typegoosev6/src/model/foo/bar.ts) ? This bar hadn't model  in master branch.

**Comment by hasezoey** [comment]
> Why did you add getModelForClass(Bar) to the nested bar ? This bar hadn't model in master branch

because i prefer
```ts
const bar = new BarModel({ fieldOne: 'fieldOne', fieldTwo: 'fieldTwo' } as Bar);

return await FooModel.create({ bar: bar } as Foo);
```
over
```ts
const bar = new Bar();
bar.fieldOne = 'fieldOne';
bar.fieldTwo = 'fieldTwo';

const foo = new Foo();
foo.bar = bar;

const fooModel: InstanceType<Foo> = new FooModel(foo);
return await fooModel.save();
```

the main difference i mean with it:
```ts
const bar = new Bar();
bar.fieldOne = 'fieldOne';
bar.fieldTwo = 'fieldTwo';
```

**Comment by fabioformosa** [comment]
Ok, test passed. 
I expected you applied the new option `customName` to one of two Bar classes or both of them, but not. 
How did you avoid conflicts on Bar class name?

**Comment by hasezoey** [comment]
[does this solve your question](https://hasezoey.github.io/typegoose/guides/advanced/models-with-same-name/#schemaoptions-collection)?

**Comment by hasezoey** [comment]
this can be closed because it will not be included in this repo anymore
(and i tried to integrate it | self-made version in v6.0.0, and failed)

@Ben305

**Pull request closed without merging**